libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary.hs

```diff
@@ -205,6 +206,14 @@ specUpgrade BinaryUpgradeOpts {isScriptUpgradeable, isTxUpgradeable} =
       specTxUpgrade @era
     when isScriptUpgradeable $
       specScriptUpgrade @era
+    -- This is a test that ensures that binary version of a TxOut is backwards compatible as it is
+    -- stored in the ledger state. This property is only important for using MemPack with UTxOHD
+    embedTripSpec
+      (eraProtVerHigh @(PreviousEra era))
+      (eraProtVerLow @era)
+      (mkTrip encodeMemPack decNoShareCBOR)
+      $ \(txOutCur :: TxOut era) (txOutPrev :: TxOut (PreviousEra era)) ->
+        upgradeTxOut txOutPrev `shouldBe` txOutCur
```

Comment:
If my understanding above is correct, I'm not sure this tests the desired property. It tests that, eg, a Babbage TxOut can be read back as an Alonzo TxOut, but not that a Conway TxOut can be read back as an Alonzo one. If this test as-is passes for all eras, then there's a good chance that what I'm saying will work, but I don't think it's guaranteed.

I feel like this test should be a loop over all previous eras, given the current era.

Reply:
> eg, a Babbage TxOut can be read back as an Alonzo TxOut

It's the opposite. `TxOut AlonzoEra` must be read as `TxOut BabbageEra`

We only need to test one era transition because we can rely on this relationship to be transitive.

If I can read `TxOut MaryEra` as `TxOut AlonzoEra` and I can read `TxOut AllegraEra` as `TxOut MaryEra`, that means I can read `TxOut AllegraEra` as `TxOut AlonzoEra`.
We do not need to test all permutations. It is really enough to test each era boundary.